internal/surfer/gcloud/utils/resource.go

```diff
@@ -0,0 +1,177 @@
+	}
+	return ""
+}
+
+// GetCollectionPathFromSegments constructs the base gcloud collection path from a
+// structured resource pattern, according to AIP-122 conventions.
+// It joins the literal collection identifiers with dots.
+// Example: `projects/{project}/locations/{location}/instances/{instance}` -> `projects.locations.instances`.
+func GetCollectionPathFromSegments(segments []api.PathSegment) string {
+	var collectionParts []string
+	for i := 0; i < len(segments)-1; i++ {
+		// A collection identifier is a literal segment followed by a variable segment.
+		if segments[i].Literal != nil && segments[i+1].Variable != nil {
+			collectionParts = append(collectionParts, *segments[i].Literal)
+		}
+	}
+	return strings.Join(collectionParts, ".")
+}
+
+// IsPrimaryResource determines if a field represents the primary resource of a method.
+func IsPrimaryResource(field *api.Field, method *api.Method) bool {
+	if method.InputType == nil {
+		return false
+	}
+	// For `Create` methods, the primary resource is identified by a field named
+	// in the format "{resource}_id" (e.g., "instance_id").
+	if strings.HasPrefix(method.Name, "Create") {
+		resourceName, err := GetResourceName(method)
+		if err == nil && resourceName != "" && field.Name == strcase.ToSnake(resourceName)+"_id" {
+			return true
+		}
+	}
+	// For `Get`, `Delete`, and `Update` methods, the primary resource is identified
+	// by a field named "name", which holds the full resource name.
+	if (strings.HasPrefix(method.Name, "Get") || strings.HasPrefix(method.Name, "Delete") || strings.HasPrefix(method.Name, "Update")) && field.Name == "name" {
+		return true
+	}
+	return false
+}
+
+// GetResourceName extracts the name of the resource from a method's input message.
+// For example, for `CreateInstanceRequest`, it would return "Instance".
+func GetResourceName(method *api.Method) (string, error) {
+	if method.InputType == nil {
+		return "", fmt.Errorf("method input type is nil")
+	}
+	for _, f := range method.InputType.Fields {
+		if msg := f.MessageType; msg != nil && msg.Resource != nil {
+			return msg.Name, nil
+		}
+	}
+	return "", fmt.Errorf("resource message not found in input type")
+}
+
+// GetResourceForMethod finds the `api.Resource` definition associated with a method.
+// This is a crucial function for linking a method to the resource it operates on.
+func GetResourceForMethod(method *api.Method, model *api.API) *api.Resource {
+	if method.InputType == nil {
+		return nil
+	}
+
+	// Strategy 1: For `Create` and `Update`, the request message usually contains
+	// a field that *is* the resource message. This message is annotated with `(google.api.resource)`.
+	for _, f := range method.InputType.Fields {
+		if msg := f.MessageType; msg != nil && msg.Resource != nil {
+			return msg.Resource
+		}
+	}
+
+	// Strategy 2: For `Get`, `Delete`, and `List`, the request message has a `name`
+	// or `parent` field with a `(google.api.resource_reference)`.
+	var resourceType string
```

Comment:
This whole block of code is hard to read as is, I would probably extract it to a helper function, then you can just return values which would read much nicer than a local plus breaks.

Reply:
I created #3363 to simplify the logic in this block, and I've added more comments to the code to help clarify how it works.